credsweeper/rules/rule.py

```diff
@@ -37,72 +42,95 @@ class Rule:
+        else:
+            self._malformed_rule_error(rule_dict, Rule.SEVERITY)
+        if rule_type := getattr(RuleType, str(rule_dict[Rule.TYPE]).upper(), None):
+            self.__rule_type: RuleType = rule_type
+        else:
+            self._malformed_rule_error(rule_dict, Rule.TYPE)
+        self.__patterns = Rule._get_patterns(self.rule_type, rule_dict[Rule.VALUES])
+        # auxiliary fields
+        self.__filters = self._get_filters(rule_dict.get(Rule.FILTER_TYPE))
+        self.__pattern_type = Rule._get_pattern_type(self.rule_type, len(self.patterns))
+        self.__use_ml = bool(rule_dict.get(Rule.USE_ML))
+        self.__validations = self._get_validations(rule_dict.get(Rule.VALIDATIONS))
+        self.__required_substrings = [i.strip().lower() for i in rule_dict.get(Rule.REQUIRED_SUBSTRINGS, [])]
+        self.__min_line_len = int(rule_dict.get(Rule.MIN_LINE_LEN, MAX_LINE_LENGTH))
+        self.__usage_list: List[str] = rule_dict.get(Rule.USAGE_LIST, [])
+
+    def _malformed_rule_error(self, rule_dict: Dict, field: str):
+        raise ValueError(f"Malformed rule '{self.__rule_name}'."
+                         f" field '{field}' has invalid value"
+                         f" '{rule_dict.get(field)}'")
+
+    @cached_property
     def rule_name(self) -> str:
         """rule_name getter"""
         return self.__rule_name
 
-    @rule_name.setter
-    def rule_name(self, rule_name: str) -> None:
-        """rule_name setter"""
-        self.__rule_name = rule_name
-
-    @property
+    @cached_property
     def rule_type(self) -> RuleType:
         """rule_type getter"""
         return self.__rule_type
 
-    @rule_type.setter
-    def rule_type(self, rule_type: RuleType) -> None:
-        """rule_type getter"""
-        self.__rule_type = rule_type
-
-    @property
+    @cached_property
     def severity(self) -> Severity:
         """severity getter"""
         return self.__severity
 
-    @severity.setter
-    def severity(self, severity: str) -> None:
-        """severity setter"""
-        if severity_obj := Severity.get(severity):
-            self.__severity = severity_obj
-        else:
-            raise ValueError(f'Malformed rule config file. Rule severity "{severity}" is invalid.')
-
-    @property
+    @cached_property
     def filters(self) -> List[Filter]:
         """filters getter"""
         return self.__filters
 
-    @filters.setter
-    def filters(self, filter_type: str) -> None:
-        """filters setter"""
-        if filter_type == "" or filter_type is None:
-            self.__filters = []
-        else:
+    def _get_filters(self, filter_type: Union[None, str, List[str]]) -> List[Filter]:
+        """
+            filter_type: str - applies Group of filter
+                         list - creates specific set of Filters
+        """
+        if not filter_type:
+            # empty line or skipped element mean empty list of filters
+            return []
```

Comment:
It seems this `if` condition can be removed.

There are 3 cases for this `if` condition.
1. `filter_type` = `None`
  - It can be handled by `raise ValueError` which in tail of function.
2. `filter_type` = `""`
  - It can be handled by `raise ValueError` which in tail of function, through the `elif isinstance(filter_type, str):` if-else chain.
3. `filter_type` = `[]`
  - It can be handled on `elif isinstance(filter_type, list):` if-else chain. It will return empty list `[]`.

Reply:
Then, it means filter_type will be mandatory  field. Some tests failed due were developed without the check.

Reply:
Yeah that's right. Thank you for a quick fix.